ocaml/xapi/bios_strings.ml

```diff
@@ -13,45 +13,130 @@
  *)
 module D = Debug.Make(struct let name="bios_strings" end)
 open D
-open Stdext.Xstringext
 
 let dmidecode_prog = "/usr/sbin/dmidecode"
 
 let remove_invisible str =
-  let l = String.split_on_char '\n' str in
-  let l = List.filter (fun s -> not (String.startswith "#" s)) l in
-  let str = String.concat "\n" l in
-  String.fold_left (fun s c -> if c >= ' ' && c <= '~' then s ^ (String.of_char c) else s) "" str
+  Astring.String.filter (fun c -> c >= ' ' && c <= '~') str
 
-(* obtain the BIOS string with the given name from dmidecode *)
-let get_bios_string name =
+(* A single record from the output of dmidecode,
+ * without its type, handle nor size.
+ * Arrays are represented by a string with values separated by eol *)
+type record = { name: string; values: (string * string) list; }
+
+(* Parser module for the output of command dmidecode -q *)
+module P = struct
+  open Angstrom
+  let is_eol =   function | '\n' -> true | _ -> false
+  let is_tab =   function | '\t' -> true | _ -> false
+  let is_space = function | ' '  -> true | _ -> false
+  let is_sep =   function | ':'  -> true | _ -> false
+
+  let space = skip is_space
+  let sep = skip is_sep
+  let tab = skip is_tab
+
+  let token = take_till is_eol <* end_of_line
+  let key = tab *> take_till is_sep <* sep
+  let array_count = space *> token
+    >>= fun str -> match int_of_string_opt str with
+    | Some n -> return (Some n)
+    | None -> fail "Couldn't parse number"
+  let v_line = tab *> tab *> token
+
+  (* Inline value example:
+   *	String 1: Dell System
+   *)
+  let inline_value =
+    lift2 (fun k v -> k, v)
+    key (* k *)
+    (space *> token) (* v *)
+
+  (* Multiple-line value examples:
+   *	Features:
+   *		Board is a hosting board
+   *		Board is replaceable
+   *
+   *	Items: 1
+   *		0x0000 (OEM-specific)
+   *)
+  let multi_value =
+    lift3 (fun k _ v_lines -> k, String.concat "\n" v_lines)
+    key (* k *)
+    ((end_of_line >>| fun _ -> None) <|> array_count) (* number of elements, ignored *)
+    (many1 v_line) (* v_lines *)
+
+  let record =
+    lift2 (fun name values -> {name; values})
+    (token) (* name *)
+    (many (multi_value <|> inline_value)) (* values *)
+
+  let record_list = sep_by1 end_of_line record
```

Comment:
Some suggestions for names:
* `record_list` -> `records`
* `v_lines` -> `values` or `items`
* `token` -> `value` or `item` (token is a very generic term inside a parser)

Reply:
I'd rather use `value_lines` rather than `values`, as I treat it as a single value.

I'm writing tests for the parser, in particular I've found out I need to fix the case where a record cannot be parsed entirely, but a partial record is generated, it would be pretty nasty to debug if it ever happened in a live system.